### PR TITLE
Fix #7611: Add support for String Literal Types in find all refs and occurances

### LINKED ISSUE #7611: Rename and fine all references of Typescript's string literal values
_From @zivni on March 20, 2016 8:40_

Type script [string literals](https://github.com/Microsoft/TypeScript/pull/5185) allow to have a code such as 
```
//interfaces.d.ts
type Options = "option 1 " | "option 2" 

//file.ts
let myOption: Options = "option 1" // "option 3" will give us an error, right in the IDE.
```

It would be nice to be able to rename and find all references of a specific value. e.g. find/rename all occurrences of "option 1". Currently we can only find the references of the type itself.

_Copied from original issue: Microsoft/vscode#4469_

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #7611

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -6151,6 +6155,52 @@ namespace ts {
...
+                const type = getStringLiteralTypeForNode(node, typeChecker);
+
+                if (!type) {
+                    // nothing to do here. moving on
+                    return undefined;
+                }
+
+                const references: ReferenceEntry[] = [];
+
+                forEach(sourceFiles, sourceFile => {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Just use a for-of

File: src/services/services.ts [hunk 2]
```diff
@@ -6151,6 +6155,52 @@ namespace ts {
...
+                    const possiblePositions = getPossibleSymbolReferencePositions(sourceFile, type.text , sourceFile.getStart(), sourceFile.getEnd());
+                    getReferencesForStringLiteralInFile(sourceFile, type, possiblePositions, references);
+                });
+
+                return [{
+                    definition: {
+                        containerKind: "",
+                        containerName: "",
+                        fileName: node.getSourceFile().fileName,
+                        kind: ScriptElementKind.variableElement,
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: variableElement?

File: src/services/services.ts [hunk 3]
```diff
@@ -6151,6 +6155,52 @@ namespace ts {
...
+                        name: type.text,
+                        textSpan: createTextSpanFromBounds(node.getStart(), node.getEnd())
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: pass sourceFile to `getStart`

File: src/services/services.ts [hunk 4]
```diff
@@ -6151,6 +6155,52 @@ namespace ts {
...
+                    },
+                    references: references
+                }];
+
+                function getReferencesForStringLiteralInFile(sourceFile: SourceFile, searchType: Type, possiblePositions: number[], references: ReferenceEntry[]): void {
+                    forEach(possiblePositions, position => {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: for-of?

File: src/services/services.ts [hunk 5]
```diff
@@ -6151,6 +6155,52 @@ namespace ts {
...
+                        cancellationToken.throwIfCancellationRequested();
+
+                        const node = getTouchingWord(sourceFile, position);
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: So this won't work within the blank space in a string literal, right?

File: tests/cases/fourslash/renameStringLiteralTypes.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/// <reference path='fourslash.ts' />
+
+
+////interface AnimationOptions {
+////    deltaX: number;
+////    deltaY: number;
+////    easing: "ease-in" | "ease-out" | "[|ease-in-out|]";
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/renameStringLiteralTypes.ts]: Would this work if you tried renaming adjacent to the dashed portions? Would it work if you had spaces around?

File: src/services/services.ts [hunk 7]
```diff
@@ -7736,6 +7808,28 @@ namespace ts {
...
+                    const sourceFile = declaration.getSourceFile();
+                    const canonicalName = getCanonicalFileName(ts.normalizePath(sourceFile.fileName));
+                    if (canonicalName === canonicalDefaultLibName) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            function createTriggerSpanForNode(node: Node) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I'd pass along the sourcefile

**Comment by DanielRosenwasser** [comment]
Can you add a test where an ordinary string literal (of type `string`) don't get affected?

File: src/services/services.ts (hunk 5 above)
Comment (mhegazy) [inline src/services/services.ts]: no it should. the literal is one token.

File: tests/cases/fourslash/renameStringLiteralTypes.ts (hunk 6 above)
Comment (mhegazy) [inline tests/cases/fourslash/renameStringLiteralTypes.ts]: yes, the whole string literal is one token we do not break it apart and scan inside it.

File: src/services/services.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Right, it's not as granular as characters, just tokens

File: src/services/services.ts (hunk 2 above)
Comment (mhegazy) [inline src/services/services.ts]: ideally a new element "stringElement", but will need some handling on the VS side, so can get that done first.

File: src/services/services.ts (hunk 3 above)
Comment (mhegazy) [inline src/services/services.ts]: more or less the same impact, i do not have the source file handy anyways.

File: src/services/services.ts (hunk 7 above)
Comment (mhegazy) [inline src/services/services.ts]: done.

**Comment by mhegazy** [comment]
@DanielRosenwasser any more comments?

File: src/services/services.ts [hunk 8]
```diff
@@ -7671,53 +7721,75 @@ namespace ts {
...
+                    node.kind === SyntaxKind.StringLiteral ||
+                    isLiteralNameOfPropertyDeclarationOrIndexAccess(node)) {
+                    const symbol = typeChecker.getSymbolAtLocation(node);
+
+                    // Only allow a symbol to be renamed if it actually has at least one declaration.
+                    if (symbol) {
+                        const declarations = symbol.getDeclarations();
+                        if (declarations && declarations.length > 0) {
+                            // Disallow rename for elements that are defined in the standard TypeScript library.
+                            if (forEach(declarations, isDefinedInLibraryFile)) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: You should just do a pass at the end for all the references you get, instead of doing this logic here and below

**Comment by DanielRosenwasser** [comment]
I meant a test in which contextually typed strings are considered references, but others of the same content are not.

Other than that and my last feedback, looks great :+1:

**Pull request merged**